### PR TITLE
Add `mz_internal.mz_cluster_replica_sizes`

### PR DESCRIPTION
**Pull request by umanwizard**
Arjun and Seth have said it's okay to expose this.

### Motivation

  * This PR adds a known-desirable feature.

    (Not sure if there's an issue, @ggnall do you know?)
    
    We want to be able to expose the percentage of memory and CPU being used. The most elegant way to do this is to expose the sizes of the replicas, and create a view on top that divides the values in `mz_internal.mz_cluster_replica_metrics` by them.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

This is tough to test because we don't have memory/cpu limits, even in Cloudtest, but I verified by hand that it works in cloud.

- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add mz_internal.mz_cluster_replica_sizes table.

### CODE REVIEW

File: src/adapter/src/catalog/builtin_table_updates.rs [hunk 1]
```diff
@@ -155,7 +170,23 @@ impl CatalogState {
...
+
+        if let (Some(processes), Some(cpu), Some(MemoryLimit(ByteSize(memory)))) =
+            (processes, cpu, memory)
+        {
+            updates.push(BuiltinTableUpdate {
+                id: self.resolve_builtin_table(&MZ_CLUSTER_REPLICA_SIZES),
+                row: Row::pack_slice(&[
+                    Datum::UInt64(id),
+                    Datum::UInt64(u64::cast_from(processes.get())),
+                    Datum::UInt64(u64::cast_from(cpu.as_millicpus()) * 1_000_000),
```
Comment (jkosh44) [inline src/adapter/src/catalog/builtin_table_updates.rs:183]: Is it possible for this to ever overflow?

File: src/adapter/src/catalog/builtin.rs [hunk 2]
```diff
@@ -1389,6 +1389,16 @@ pub static MZ_CLUSTER_REPLICA_STATUSES: Lazy<BuiltinTable> = Lazy::new(|| Builti
...
 
+pub static MZ_CLUSTER_REPLICA_SIZES: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
+    name: "mz_cluster_replica_sizes",
+    schema: MZ_INTERNAL_SCHEMA,
+    desc: RelationDesc::empty()
+        .with_column("replica_id", ScalarType::UInt64.nullable(false))
+        .with_column("processes", ScalarType::UInt64.nullable(false))
+        .with_column("cpu_nano_cores", ScalarType::UInt64.nullable(false))
+        .with_column("memory_bytes", ScalarType::UInt64.nullable(false)),
+});
```
Comment (jkosh44) [inline src/adapter/src/catalog/builtin.rs:1400]: We should add this to the docs: https://materialize.com/docs/sql/system-catalog/mz_internal/

**Review by jkosh44 — APPROVED** [review]
Code LGTM. I had a couple of questions about the design, though these may have already been discussed.
- Is there a reason that we want a new table `mz_cluster_replica_sizes`, instead of just adding columns to `mz_cluster_replicas`?
- If the values in `mz_cluster_replica_sizes` can be derived from the replica size, would it make more sense to have a size to (process, cpu, memory) mapping table? Then we could implement `mz_cluster_replica_sizes` as a view with a join over that mapping table and `mz_cluster_replicas`.

**Comment by jkosh44** [comment]
Also we should add a test that creates a replica and queries the new builtin table.

File: src/adapter/src/catalog/builtin_table_updates.rs (hunk 1 above)
Comment (umanwizard) [reply src/adapter/src/catalog/builtin_table_updates.rs:183]: No, it will generally be less than 100 billion.

**Comment by umanwizard** [comment]
I spoke offline with Joe about the issues he raised. We are not documenting this, nor putting it in the existing `mz_cluster_replicas` table, because we don't intend for it to be publicly used (though it's not secret, and fine if users discover it)

**Comment by benesch** [comment]
Dropping in a few post-merge thoughts.

  * Really excited we're finally past the logjam of how to expose this stuff without committing to it.
  * I would like to discuss the design of this table—specifically the question @jkosh44 raised. I was envisioning this as a table with columns `(size, cpu_nano_cores, memory_bytes, scale)`—i.e., mapping size names to resource allocations. That has a benefit over this design: you don't actually need to turn on a replica at a given size to know how big it is. I think that'll be important for some downstream use cases. If we're going to expose this information, I think we might as well do it in the most ergonomic way possible, rather than making users jump through hoops if they want to know how big an e.g. 2xlarge is without actually making one.
  * We should be in the habit of testing this stuff, even when it's hard. We can always engage @philip-stoev for advice/support when the testing gets hard.
  * As discussed with @sjwiesman: we should document the existence of this table in the `mz_internal` docs, though with a warning that tells users not to rely on the values of `cpu_nano_cores` and `memory_bytes` being anything in particular.

**Comment by umanwizard** [comment]
I have a follow-up PR for docs and tests in the works.

As for whether to have an entry per replica or a static table with size mappings — I actually discussed that with Joe, sorry I neglected to document it here. I thought it was better to do the former, because in the future, we might want to change the size mappings for new replicas, without changing it for old ones. The current approach lets is preserve maximum flexibility.

Do you think it’s certain that we won’t want to do that?

**Pull request merged**